### DISCUSSION TITLE
Resize file in directory keep file name

### DISCUSSION THREAD

**Discussion by catman78** [ImageMagick/ImageMagick#2790] [Help]
Hello all,

I'm a new user to Imagemagick.
I have a powershell script that will be using FileSystemWatcher to monito for newly created .jpegs.
As soon as file is created I need to:

1. copy image to backup directory
2. resize the image and keep the same file name in current directory.

How can I achieve step 2 without renaming file?

Thanks

**Comment by snibgo** [comment]
What version of IM?

The usual "magick" command needs input and output filenames, and these can be the same. For example:
```
magick myfile.jpg -resize 400x400 myfile.jpg
```

**Comment by catman78** [comment]
Thanks version 7.  It works however FileSystemWatcher thinks it's a new file and get's into loop.... Ugh

**Reply by t246246** [reply]
In general, you have to open file read/write mode.
Open file, read the whole content in memory, send it imagemagick on another process from STDIN.
Get the result from imageMagick STDOUT and receive it.
Then seek the first position to the file and write the content, close the file.

You need to know some knowledge of PIPE or SOCKET.

--- EDIT
To illustrate it, here is a sample program written in perl, which works fine in Unix-friendly environment, but might not work on Windows. (It should work in MSYS2 perl and magick.)  I use Open2 for communication - ruby, or python have the same concept. 

```
#!/usr/bin/perl -w

use strict;
use IPC::Open2;

my $content;
open(my $fh, "+<", $ARGV[0]) or die;
{
    local($/) = undef;
    $content = <$fh>;
}

my $pid = open2(my $cout, my $cin, "/mingw64/bin/magick - -resize 100x100! JPEG:-") or die;
print $cin $content;
close($cin);
{
    local($/) = undef;
    $content = <$cout>;
}
close($cout);
waitpid($pid, 0);

truncate($fh, 0);
seek($fh, 0, 0);
print $fh $content;
close($fh);

```
save this as update.pl and do like this:
$ magick logo: logo.jpg
$ magick logo.jpg info:-
logo.jpg JPEG 640x480 640x480+0+0 8-bit sRGB 60304B 0.047u 0:00.083
$ perl ./update.pl logo.jpg
$ magick logo.jpg info:-
logo.jpg JPEG 100x100 100x100+0+0 8-bit sRGB 5333B 0.031u 0:00.069

--- EDIT END


Or if you can create a HTTP service  which receive image as a POST and return the resized image whose Content-Type is image/jpeg, that would be easier to understand these days.

**Reply by catman78** [reply]
Thanks for help.  Well I have no knowledge of that...Where can I find info about doing that?

**Reply by t246246** [reply]
Well, it is not so easy.  If you don't mind to create temporary file in another directory (that are not watched),

create resized jpeg in another place.
open original file in read and write mode and truncate it.
read from resized jpeg and write the content to original file.

I'm not familiar with powershell but it has [System.io.File]::Open, so I hope that may be your help.  See https://stackoverflow.com/questions/26156715/powershell-open-a-file-for-write-only-for-lock-and-doing-touch-so-to-change for example.  Of course you can write it in another language (I hope that is callable from powershell.)

**Reply by catman78** [reply]
Hey thanks again for your suggestions.  Will post when I have solution.

**Reply by t246246** [reply]
In Windows environment, I cannot find good scripting language, so to use Magick.NET is easier solution.
```
using ImageMagick;
using System;
using System.IO;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            using (FileStream f = File.Open(args[0], FileMode.Open, FileAccess.ReadWrite))
            {
                using (MemoryStream m = new MemoryStream())
                {
                    f.CopyTo(m);
                    m.Seek(0L, SeekOrigin.Begin);
                    using (MagickImage img = new MagickImage(m))
                    {
                        img.Resize(new MagickGeometry(args[1]));
                        f.SetLength(0L);
                        f.Seek(0L, SeekOrigin.Begin);
                        img.Write(f);
                    }
                }
            }
        }
    }
}
```
Compile it to a .exe, and take the **full path** of your file to be resized as a 1st argument, and resize parameter to ImageMagick (e.g. 100x100!) as a 2nd argument.

I checked `FileSystemWatcher` and it does not raise `Created` event for files that are only modified.
Yon can check it to add some code.  This will raise `Created` event once for randomly generated file.
For modified image file given as args[0], only `Changed` event raised twice, one for truncate (SetLength()), one for write.

```
using ImageMagick;
using System;
using System.IO;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            string dir = Path.GetDirectoryName(args[0]);
            string dummy;
            do
            {
                dummy = Path.Combine(dir, Path.GetRandomFileName());
            } while (File.Exists(dummy));

            try
            {
                using (FileSystemWatcher w = new FileSystemWatcher())
                {
                    w.Path = dir;
                    w.Created += OnCreated;
                    w.Changed += OnChanged;
                    w.EnableRaisingEvents = true;

                    using (File.Create(dummy)) { };

                    using (FileStream f = File.Open(args[0], FileMode.Open, FileAccess.ReadWrite))
                    {
                        using (MemoryStream m = new MemoryStream())
                        {
                            f.CopyTo(m);
                            m.Seek(0L, SeekOrigin.Begin);
                            using (MagickImage img = new MagickImage(m))
                            {
                                img.Resize(new MagickGeometry(args[1]));
                                f.SetLength(0L);
                                f.Seek(0L, SeekOrigin.Begin);
                                img.Write(f);
                            }
                        }
                    }
                }
            }
            finally
            {
                File.Delete(dummy);
            }
        }

        private static void OnCreated(object source, FileSystemEventArgs e) =>
            Console.WriteLine($"File: {e.FullPath} {e.ChangeType}");
        private static void OnChanged(object source, FileSystemEventArgs e) =>
            Console.WriteLine($"File: {e.FullPath} {e.ChangeType}");
    }
}
```

As you see C# has `SystemFileWatcher`, you can easily rewrite your whole Powershell script :)

**Reply by catman78** [reply]
Thanks for writing that up.  I have no experience with C# and not sure what the code is doing. 

I have this working.If($FileWatcher){$FileWatcher.Dispose();$FileWatcher = $null}

$FileWatcher = New-Object System.IO.FileSystemWatcher -Property @{
 
    NotifyFilter = [IO.NotifyFilters]'FileName, LastWrite'
    Path = 'C:\ImageTest\Source\'

}
 
Register-ObjectEvent -InputObject $FileWatcher -EventName Created -Action {

    $BackupPath = 'C:\ImageTest\Backup'
    $ResizePath = 'C:\ImageTest\Resize'
    $SourcePath = 'C:\ImageTest\Source'

    #Path to magick.exe
    $im_exe = "C:\Program Files\ImageMagick-7.0.10-Q16-HDRI\magick.exe"
	
    #get extension type
    $extn = [IO.Path]::GetExtension($event.SourceEventArgs.FullPath)
	
	#get file name
	$filename = $event.SourceEventArgs.Name
 
    #look for jpeg and .jpg only
    If ($extn -match 'jpg|jpeg'){
   
        try{     
        #copy original file to Backup Folder
        #Copy-Item -Path $event.SourceEventArgs.FullPath -Destination $BackupPath -ErrorAction Stop}
        Copy-Item -Path $event.SourceEventArgs.FullPat -Destination $BackupPath -ErrorAction Stop}
        catch{ $ErrorMessage = $_.Exception.Message
               $FailedItem = $_.Exception.ItemName 
               
               $ErrorMessage | Out-File C:\ImageTest\error.log -Append
               break}

        Write-Host "Backup " $BackupPath\$($event.SourceEventArgs.Name) 

        #call imgagemagick pass file to process then move resized file to Resize folder
        &$im_exe $event.SourceEventArgs.FullPath -resize 50% "$ResizePath\$($event.SourceEventArgs.Name)"
        
        Write-Host "Resize " $event.SourceEventArgs.FullPath 

        #replace contents of orginal file in Source with resize file
        Get-Content -path $ResizePath\$($event.SourceEventArgs.Name) -raw | Set-Content -path $SourcePath\$($event.SourceEventArgs.Name)

        Write-Host "Replace " $event.SourceEventArgs.FullPath

            	
		#get file to delete and dispose to release
        $newstreamreader = New-Object System.IO.StreamReader("$ResizePath\$($event.SourceEventArgs.Name)")
        $newstreamreader.Dispose()

        #delete file in resize folder
		Remove-Item $ResizePath\$($event.SourceEventArgs.Name)

        Write-Host "Delete from Resize Folder"
    }
 
}
